### PR TITLE
health_checks: add option to always log successful health checks

### LINKED ISSUE #32119: Missing health check logs
I'm testing a scenario where I have a ``STRICT_DNS`` cluster with a single endpoint. The DNS resolution of that endpoint returns 5 different IPs. I enabled health checks for that clusters as well as health checks logging to a file.
In my setup, only 1 out of the 5 IPs will respond to the health check, so it's expected that 5 simultaneous health checks will be performed on each interval, 1 of them will succeed and the rest won't, and that this will be indicated in the health check logs.
However, what I observe is that in the first interval, 5 log lines are printed, describing the health checks correctly, but for later intervals, only the 4 health checks that are failing are being logged, while the health check that is passing is discarded from the log (although I can verify that the help check is actually executing on that IP as well).

Config:
```
static_resources:
  listeners:
  - name: main
    stat_prefix: main
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 1234
    filter_chains:
    - name: main
      filters:
      - name: tcp_proxy
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
          stat_prefix: main
          cluster: main_cluster
          max_connect_attempts: 1

  clusters:
  - name: main_cluster
    connect_timeout: 1s
    type: STRICT_DNS
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: main_cluster
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: myhost.com
                port_value: 1111
    health_checks:
      timeout: 1s
      interval: 10s
      no_traffic_interval: 10s
      no_traffic_healthy_interval: 10s
      unhealthy_threshold: 1
      healthy_threshold: 1
      reuse_connection: false
      tcp_health_check:
        send:
          text: 6865616C7468
        receive:
        - text: 4563686F206865616C7468
      event_logger:
      - name: logger
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.health_check.event_sinks.file.v3.HealthCheckEventFileSink
          event_log_path: /data/health.json
      always_log_health_check_failures: true
```

Health check log output:
```
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.1","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:00:40.811Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":true},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.3","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","add_healthy_event":{"first_check":true},"timestamp":"2024-01-31T12:00:40.817Z","metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.5","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:00:41.815Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":true},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.2","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:00:41.815Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":true},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.4","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:00:41.815Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":true},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.1","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:00:50.806Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.5","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:00:52.819Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.4","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:00:52.820Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.2","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:00:52.820Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.1","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:01:00.812Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.5","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:01:03.821Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.4","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:01:03.822Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.2","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:01:03.822Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.1","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:01:10.804Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.2","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:01:14.813Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.5","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:01:14.813Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"TCP","host":{"socket_address":{"protocol":"TCP","address":"10.5.0.4","port_value":5556,"resolver_name":"","ipv4_compat":false}},"cluster_name":"main_cluster","timestamp":"2024-01-31T12:01:14.813Z","health_check_failure_event":{"failure_type":"NETWORK_TIMEOUT","first_check":false},"metadata":{"filter_metadata":{},"typed_filter_metadata":{}},"locality":{"region":"","zone":"","sub_zone":""}}
```

Note that the health check for ``10.5.0.3`` (the only responding IP) is only logged once at the beginning.

### PR DESCRIPTION
**Pull request by ohadvano**
Additional Description: Resolves #32119. This allows the option to always log successful health checks. On the first successful health check, only ``logAddHealthy`` is called. On consecutive successful health checks, ``logSuccessfulHealthCheck`` is called.
Risk Level: low (config guarded)
Testing: unit tests
Docs Changes: API docs
Release Notes: none
Platform Specific Features: none

### CODE REVIEW

**Comment by KBaichoo** [comment]
/assign @botengyao 

PTAL on API @lizan

File: source/common/upstream/health_checker_event_logger.cc [hunk 1]
```diff
@@ -37,6 +37,13 @@ void HealthCheckEventLoggerImpl::logAddHealthy(
   });
 }
 
+void HealthCheckEventLoggerImpl::logSuccessfulHealthCheck(
+    envoy::data::core::v3::HealthCheckerType health_checker_type,
+    const HostDescriptionConstSharedPtr& host) {
+  createHealthCheckEvent(health_checker_type, *host,
+                         [](auto& event) { event.mutable_successful_health_check_event(); });
```
Comment (botengyao) [inline source/common/upstream/health_checker_event_logger.cc:44]: This looks like a no-op, and what is this `mutable_successful_health_check_event` for?

**Review by botengyao — COMMENTED** [review]
Thanks for adding this feature!

File: api/envoy/data/core/v3/health_check_event.proto [hunk 2]
```diff
@@ -55,6 +55,12 @@ message HealthCheckEvent {
     // Host addition.
     HealthCheckAddHealthy add_healthy_event = 5;
 
+    // A health check was successful. Note: a host will be considered healthy either if it is
+    // the first ever health check, or if the healthy threshold is  reached. This kind of event
```
[the remark this answers is not in the document]
Comment (ohadvano) [reply api/envoy/data/core/v3/health_check_event.proto]: done, thanks

File: source/common/upstream/health_checker_event_logger.cc (hunk 1 above)
Comment (ohadvano) [reply source/common/upstream/health_checker_event_logger.cc:44]: when calling ``mutable_x`` it sets the kind of the event type to this (similar as the ``mutable_degraded_healthy_host`` below). I intentionally created another kind of event type, to be able to distinguish between ``add_healthy`` and ``successful_health_check``.

**Review by lizan — COMMENTED** [review]
/lgtm api

File: source/common/upstream/health_checker_event_logger.cc (hunk 1 above)
Comment (botengyao) [reply source/common/upstream/health_checker_event_logger.cc:44]: Thanks, could you add a test to cover this and verify the output similar [as this one](https://github.com/envoyproxy/envoy/blob/8318716d9aedfc6277cd605a41b606a86f3feb52/test/common/upstream/health_checker_impl_test.cc#L6468)?

**Review by botengyao — COMMENTED** [review]
LGTM module one test. Thanks!

/wait

**Review by botengyao — APPROVED** [review]
sry, could you also add a release note in current.yaml? Otherwise LGTM, Thanks!

/assign-from  @envoyproxy/senior-maintainers

**Comment by ohadvano** [comment]
Done. @lizan can you PTAL?

**Review by lizan — APPROVED** [review]

**Pull request merged**